### ISSUE TITLE
Ability to use wildcards in mocks / mock a range of requests

### ISSUE THREAD

**Issue by dominik-fielmann** [apollographql/apollo-feature-requests#206]
## Motivation
I am currently developing a prototype / proof of concept for an app where the user searches for an id and gets a view of the product. I want to prepare the GraphQL integration, but for now I want to mock all requests to the backend and return default dummy data to show to the user.

## The problem
To use the MockedProvider to mock away the graphql backend, I have to specify exactly which variables the query will be called with. But I have no way to anticipate what the user will try to search for.

## Feature Request
Make it possible to use wildcards in mocks or otherwise leave open a part of the query, so it matches regardless of a variable value.

## Example / Proposed feature

This matches only one very specific query:
```
const mocks = [
  {
    request: {
      query: GET_ARTICLE,
      variables: { articleNumber: "123456" }
    },
    result: {
      data: { article: testArticle } 
    }
  }
];
```

Something like this could match all queries to GET_ARTICLE and always return the same result:

```
const mocks = [
  {
    request: {
      query: GET_ARTICLE
    },
    result: {
      data: { article: testArticle } 
    }
  }
];
```
Alternatively:
``` 
    request: {
      query: GET_ARTICLE
      variables: { articleNumber: ANY_VALUE }
    }
```
where ANY_VALUE would be a wildcard constant that Apollo recognises.

**Comment by dylanwulf** [comment]
I prefer to avoid MockedProvider entirely and instead go with the approach outlined here: https://medium.com/free-code-camp/a-new-approach-to-mocking-graphql-data-1ef49de3d491

This is the approach I use in all my client-side tests. It automatically provides mock data of the correct type, but you can still specify your own mock data if you wish. You can also choose to only provide some mock data, and allow the rest of the query data to be filled automatically. 
The query arguments are provided to your mock function in the second argument, so you can choose to return different mock data depending on the query arguments (or just ignore the variables entirely)

**Comment by insidewhy** [comment]
@dominik-fielmann 

```typescript
import { MockLink, MockedResponse } from "@apollo/react-testing";
import { Operation, FetchResult, Observable } from "apollo-link";
import { addTypenameToDocument } from "apollo-utilities";
import { print, DocumentNode } from "graphql";

export interface WildcardMockedResponse extends MockedResponse {
  nMatches?: number;
}

interface StoredMock {
  result: FetchResult | (() => FetchResult) | undefined;
  // by default a mock will match an infinite number of requests
  nMatches: number;
}

/**
 * Extends MockLink to provide the ability to match request queries independent
 * of their variables and have them match 1 or more responses.
 */
export class WildcardMockLink extends MockLink {
  private wildcardMatches = new Map<string, StoredMock[]>();

  request(op: Operation): Observable<FetchResult> | null {
    const mockKey = print(op.query);
    const storedMocks = this.wildcardMatches.get(mockKey);

    if (storedMocks) {
      const nextStoredMock = storedMocks[0];

      if (--nextStoredMock.nMatches === 0) {
        storedMocks.shift();
        if (!storedMocks.length) {
          this.wildcardMatches.delete(mockKey);
        }
      }

      return new Observable(observer => {
        const { result } = nextStoredMock;
        if (result) {
          setTimeout(() => {
            observer.next(typeof result === "function" ? result() : result);
            observer.complete();
          }, 0);
        }
      });
    } else {
      return super.request(op);
    }
  }

  addWildcardMockedResponse(response: WildcardMockedResponse): void {
    const mockKey = this.buildCacheKey(response.request.query);
    const storedMocks = this.wildcardMatches.get(mockKey);
    const storedMock = {
      result: response.result,
      nMatches: response.nMatches || Number.POSITIVE_INFINITY,
    };
    if (storedMocks) {
      storedMocks.push(storedMock);
    } else {
      this.wildcardMatches.set(mockKey, [storedMock]);
    }
  }

  removeWildcardMockedResponse(request: DocumentNode): void {
    this.wildcardMatches.delete(this.buildCacheKey(request));
  }

  private buildCacheKey(request: DocumentNode): string {
    return print(this.addTypename ? addTypenameToDocument(request) : request);
  }
}
```

Then use like:

```typescript
const renderStuff = () => {
  const link = new WildcardMockLink(mocks);
  return {
    element: (
      <MockedProvider link={link}>renderMyComponent()</MockedProvider>
    ),
    link,
  };
}
```

Then use this to grab the link and the element at the same time, add wildcard responses dynamically etc. etc.

**Comment by insidewhy** [comment]
I'm open sourcing this soon, actually a much better version which also supports subscriptions and things like that :D

**Comment by insidewhy** [comment]
Released this package which supports wildcards in mocks and a bunch of other nice functionality to help writing tests: https://github.com/insidewhy/wildcard-mock-link

Is documented/tested.

**Comment by richardwu** [comment]
👍 on this

**Comment by insidewhy** [comment]
@richardwu you should use reactions instead of comments saying +1 which notify all thread watchers. At any rate wildcard-mock-link already supports this together with all other features of the mock link.

**Comment by david-morris** [comment]
@insidewhy `wildcard-mock-link` does not support jest `act`, nor does it support Apollo Client Addon for Storybook.  Having this functionality inside the apollo MockedProvider would greatly speed up our tests and improve our stories.

Please do not dismiss this use case as solved because there is a workaround.

**Comment by insidewhy** [comment]
> does not support jest act

It does, you can even pass `act` as an option to avoid having to use it manually if you want.

> nor does it support Apollo Client Addon for Storybook

Could you give some details on this?

**Comment by david-morris** [comment]
@insidewhy 
> you can even pass `act` as a constructor argument

Wow, thank you.  I can't believe I missed this after how many times I read the documentation. You just saved my team and tests a lot of time.

> Apollo Client Addon details

https://storybook.js.org/addons/storybook-addon-apollo-client

I was hoping to use that as a way of showing parameters of mocked mutations in the stories.  I think I can probably hack around it by processing the request callback of my mocks.

**Comment by riezebosch** [comment]
> @insidewhy
> 
> > you can even pass `act` as a constructor argument
> 
> Wow, thank you. I can't believe I missed this after how many times I read the documentation. You just saved my team and tests a lot of time.
> 
> > Apollo Client Addon details
> 
> https://storybook.js.org/addons/storybook-addon-apollo-client
> 
> I was hoping to use that as a way of showing parameters of mocked mutations in the stories. I think I can probably hack around it by processing the request callback of my mocks.

Thanks for the link to the addon! I combined it with the Play function to specifically enter data for the variables used on the mock and tell that as a distinct story: https://storybook.js.org/docs/react/writing-stories/play-function.

**Comment by RobinTail** [comment]
- https://github.com/apollographql/apollo-client/pull/6701

**Comment by jerelmiller** [comment]
Thanks for the nudge @RobinTail!

That change will be coming in the next minor 3.9.0 which should release within the next couple weeks. As such, I'm going to go ahead and close out this issue. Feel free to open a new feature request if you need further assistance. Looking forward to getting this one out!

**Comment by ARkrOSClou** [comment]
> * [Ability to dynamically match mocks apollo-client#6701](https://github.com/apollographql/apollo-client/pull/6701)

unfortunately this solution only works for 1 call, has anyone else tested this?

**Comment by jerelmiller** [comment]
Hey @ARkrOSClou 👋 

I think you'll find https://github.com/apollographql/apollo-client/pull/11178 useful, which will be released in 3.9 along with the change to dynamically match variables. You can set a `maxUsageCount` to tell `MockedProvider` to reuse a mock as many times as you need to. 

Check out the [preview docs](https://deploy-preview-11220--apollo-client-docs.netlify.app/development-testing/testing#reusing-mocks) if you'd like to try this feature out in `3.9.0-rc.1` before our public release of 3.9 🙂

**Comment by ferrata** [comment]
> Hey @ARkrOSClou 👋
> 
> I think you'll find [apollographql/apollo-client#11178](https://github.com/apollographql/apollo-client/pull/11178) useful, which will be released in 3.9 along with the change to dynamically match variables. You can set a `maxUsageCount` to tell `MockedProvider` to reuse a mock as many times as you need to.
> 
> Check out the [preview docs](https://deploy-preview-11220--apollo-client-docs.netlify.app/development-testing/testing#reusing-mocks) if you'd like to try this feature out in `3.9.0-rc.1` before our public release of 3.9 🙂

Not sure if I understand it correctly, but this code throws errors on all subsequent calls:

```ts
const mocks = [ {
    request: {
      query: ...,
    },
    maxUsageCount: Number.POSITIVE_INFINITY,
    variableMatcher: () => true,
    result: { ... }
} ]

// ... 
<MockedProvider mocks={mocks}>
// ...
```

![CleanShot 2024-05-17 at 16 28 27@2x](https://github.com/apollographql/apollo-feature-requests/assets/31716295/75abea0d-8899-4088-b6f3-3fcc06de8485)
